### PR TITLE
Impl From between Option<NonZeroInt> <-> Int

### PR DESCRIPTION
**Pull request by Riey**
Those types have same memory layout and often converted each other
but conversion is little verbose

```rust

let a: Option<NonZeroU32> = NonZeroU32::new(0);
let b: u32 = 0;

// Before
fn foo(n: Option<NonZeroU32>);
fn bar(n: u32);

foo(a);
foo(NonZeroU32::new(b));
bar(a.map_or(0, |a| a.get()));
bar(b);

// After
fn foo(n: impl Into<Option<NonZeroU32>>);
fn bar(n: impl Into<u32>);

foo(a);
foo(b);
bar(a);
bar(b);
```

### CODE REVIEW

**Comment by timvermeulen** [comment]
> ```rust
> bar(a.map(|a| a.get()).unwrap_or(0));
> ```

`bar(a.map_or(0, |a| a.get()))` is slightly less verbose.

**Comment by Riey** [comment]
@timvermeulen Thanks you for suggestion your version is better
but I think Into/From is more straightforward and doesn't need hard coded number

File: src/libcore/num/mod.rs [hunk 1]
```diff
@@ -100,6 +100,28 @@ assert_eq!(size_of::<Option<core::num::", stringify!($Ty), ">>(), size_of::<", s
                 }
             }
 
+            #[stable(feature = "from_nonzeroopt", since = "1.43.0")]
+            impl From<Option<$Ty>> for $Int {
+                doc_comment! {
+                    concat!(
+"Converts an `Option<", stringify!($Ty), ">` into an `", stringify!($Int), "`"),
+                    fn from(nonzero: Option<$Ty>) -> Self {
+                        nonzero.map($Ty::get).unwrap_or(0)
```
Comment (hellow554) [inline src/libcore/num/mod.rs]: As @timvermeulen said, this should be `nonzero.map_or(0, $Ty::get)`

**Comment by hellow554** [comment]
Some testcases would be nice. Take a look at https://github.com/rust-lang/rust/blob/master/src/libcore/tests/nonzero.rs

File: src/libcore/num/mod.rs (hunk 1 above)
Comment (Riey) [reply src/libcore/num/mod.rs]: Ok I fixed it

**Comment by hellow554** [comment]
ping @shepmaster :)

**Comment by sfackler** [comment]
@rfcbot fcp merge

**Comment by Amanieu** [comment]
I'm a bit concerned by the `Option<NonZeroInt>` to `Int` conversion, it feels a bit too implicit in that it picks 0 as the default value for a `None`. I feel that this might become a potential footgun because `None` might mean something completely different than just 0 (for example "value is missing") and the `impl Into` conversion would silently hide this.

@rfcbot concern too-much-magic

I don't have any objection for the `Int` to `Option<NonZero>` conversion though.

**Comment by SimonSapin** [comment]
I agree with @Amanieu. Just because memory representation happens to match doesn’t mean that this implementation detail always translates to the desired semantics. I feel this conversion does not belong with something as general purpose as the `From` trait.

This is not just a concern that can be resolved. If there wasn’t already an ongoing proposed FCP I would propose FCP to close.

For cases where it is desired, `.unwrap_or(0)` is easy to write and should also optimize to nothing.

> I don't have any objection for the `Int` to `Option<NonZero>` conversion though.

This already exists and is spelled `NonZero::new`

**Comment by SimonSapin** [comment]
> For cases where it is desired, `.unwrap_or(0)` is easy to write and should also optimize to nothing.

Ok it’s slightly longer, but still:

```rust
pub fn foo(x: Option<std::num::NonZeroU32>) -> u32 {
    x.map_or(0, |n| n.get())
}
```
https://rust.godbolt.org/z/6Fq_VR
```asm
example::foo:
        mov     eax, edi
        ret
```

**Comment by Riey** [comment]
@Amanieu I think Option<NonZero>::None already has zero

**Comment by Riey** [comment]
Sorry I miss clicked close button

**Comment by Riey** [comment]
@Amanieu I think Option<NonZeroInt>::None already has same meaning of 0

NonZeroInt::new already doing this

**Comment by Riey** [comment]
@SimonSapin I think implement `From` trait mean it can always converted from that type

IpAddrV4 impl From<[u8; 4]> and i32 impl From<bool> these types don't have any similar sementic but it has

**Comment by jhpratt** [comment]
> @Amanieu I think `Option::<NonZeroInt>::None` already has same meaning with 0
> 
> NonZeroInt::new already doing this

It may have the same _memory representation_, but it's far from the same semantics. I find this to be very odd behavior, so I am opposed to this for reasons already stated.

**Comment by Riey** [comment]
@jhpratt Thank you that's good point so I agree Option<NonZeroInt> doesn't have save sementic with Int

However `NonZeroInt::new` doing convert `Int -> Option<NonZeroInt>` so I think `impl From<Int> for Option<NonZeroInt>` can be possible
but as other concern `Option<NonZeroInt> -> Int` does more magic and there is no possitive opinion so I close this

**Pull request closed without merging**